documentation/ArrayExprLoweringToFIR.md

```diff
@@ -0,0 +1,303 @@
+#### FIR operations to express array operations
+In FIR, affine expressions can be implemented using the `fir.loop`, `fir.where`, `fir.extract_value` and `fir.coordinate_of` operations in combinations with the `fir.array<>` type and the `fir.box<>` type that can hold affine maps encoding linear access patterns (see [FIR  Language reference](documentation/FIRLangRef.md)).
+#### Views over array in FIR
+In FIR, one can avoid copying array sections by creating `fir.box` with that holds the memory reference to the base variable as well as the subscript describing the re-mapping.
+For instance, `A(10:20:2, :)` where `A` is `REAL, DIMENSION(20, 10)` can be written in FIR as:
+```
+//%a is the memory reference to A storage, it is of type !fir.ref<fir.array<20:5:fir.real<4>>>
+%0 = fir.gendims 2*m+1, 4*m, 2
+%1 = fir.gendims 1, n, 1
+%asection = fir.embox %a, %0, %1
+// %asection has type !fir.box<fir.ref<fir.array<5:10:fir.real<4 >>>>
+```
+Then the element (i, j) of `A(10:20:2, :)`  can be accessed with:
+```
+// fir.extract_value takes care of applying the index mapping of the array section over A.
+%asectionij = fir.extract_value %asection, %i, %j
+// % asectionij has type !fir.real<4>
+```  
+No copy into a temporary was needed to implement the example above. `fir.box` allows one to represent array and manipulate subsections in FIR without materializing them. It can be seen as if we had created a Fortran pointer over `A` (except such object would have had the type `%fir.box<fir.ptr< fir.array<5:10:fir.real<4 >>>`, notice the `fir.ptr` instead of `fir.ref`).
+
+One notable difference between using `fir.box` and materializing a section inside a temporary is that its value its not “fixed”, the evaluation of the section is delayed until it is needed. This introduce the risk that `A` variable is modified in-between the array emboxing and its usage. In the context of expression evaluation, the only thing that could modify another variable appearing inside the expression evaluation are function call (potentially hidden in user defined operators).
+Fortran Standard prohibit this by stating in section 10.1.4 point 2
+> [..] Except in those cases: the evaluation of a function reference shall neither affect nor be affected by the evaluation of any other entity within the statement
+
+The cases mentioned do not apply inside an expression. However, it is in general unsafe to propagate such box outside of an expression evaluation. This means that if `A(10:20:2, :)` if a top level expression, it is probably safer to materialize into a temporary storage and leave later optimizations determine if it is safe to keep skip the materialization.
+This means for the Fortan code:
+```
+B = A(1:10, :)
+PRINT *m, B
+```
+The bridge to FIR will not replace refences to `B` by the `fir.box` encoding ` A(1:10, :)` because it cannot ensure `A` was not modified in-between without an analysis of A potential definition in-between that does not belong to the bridge towards FIR.
```

Comment:
OK, I have reworded to:
"The bridge to FIR will not replace the reference to `B` inside the `PRINT` statement by a `fir.box` encoding ` A(1:10, :)`. It would be correct here, but dangerous in general. The reason is that the value represented by this `fir.box` is not constant. This value is the one of `A(1:10,:)` at the moment the `fir.box` is accessed with `fir.extract_value`. Emboxing an array section has the side effect of delaying its evaluation. 
In general, the bridge cannot easily ensure `A` was not modified in-between `B=A(1:10, :)` and `PRINT *m, B`. It would require an analysis of `A` variable definitions that do not belong to the bridge towards FIR."